base/array.jl

```diff
@@ -3026,3 +3026,31 @@ intersect(r::AbstractRange, v::AbstractVector) = intersect(v, r)
         _getindex(v, i)
     end
 end
+
+"""
+    wrap(Array, m::Union{Memory{T}, MemoryRef{T}}, dims)
+
+Create an array of size `dims` using `m` as the underlying memory. This can be thought of as a safe version
+of [`unsafe_wrap`](@ref) utilizing `Memory` or `MemoryRef` instead of raw pointers.
+"""
+@propagate_inbounds function wrap(::Type{Array}, m::MemoryRef{T}, dims::NTuple{N, Integer}) where {T, N}
+    len = length(m.mem)
```

Comment:
You need to call `jl_genericmemory_slice` here in some cases to make the length valid. Take a look at `reshape` for arrays for how to do that.

Also, should this just be a method of `reshape` as well? It seems to pretty exactly follow that existing API

Reply:
If it has a different start or end, then it does have a different shape though. But normally reshape doesn't allow slices like that directly (only implicitly via resize first). So I acknowledge it is not exactly identical